### ISSUE TITLE
Recursive functions aren't picked up by no-unused-vars

### ISSUE THREAD

**Issue by tav** [eslint/eslint#2095]
Given the following source:

```js
(function() {

  function fib(n) {
    if (n === 0) {
      return 0;
    }
    if (n === 1) {
      return 1;
    }
    return fib(n - 1) + fib(n - 2);
  }

  function foo(n) {
    return n;
  }

})();
```

ESLint correctly picks up that `foo` is unused, but fails to detect that `fib` is unused since it recursively refers to itself.

**Comment by nzakas** [comment]
Cab you please share the exact console output you're seeing?

**Comment by nzakas** [comment]
From a cursory look, this is behaving correctly. "used" means there's a reference to the identifier, and there is.

**Comment by tav** [comment]
Here is the console output:

```
test.js
  13:11  error  foo is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

**Comment by tav** [comment]
Here's the console output for `jshint`:

```
test.js

  line 3, char 12: 'fib' is defined but never used.
  function fib(n) {

  line 13, char 12: 'foo' is defined but never used.
  function foo(n) {

✘  FAIL · 2 Warnings
```

**Comment by tav** [comment]
As to it "behaving correctly", I guess that depends on your interpretation :wink:

I've historically used it in other tools like `'jshint` to help identify dead/cruft code which aren't used elsewhere in the file. In an ideal world, it would provide greater static analysis (e.g. detecting unused reassignments, etc.), but I'd be happy to settle for "unused outside of definition" ...

**Comment by nzakas** [comment]
Does JSHint flag `foo` as unused?

**Comment by tav** [comment]
Yes. See my earlier paste of the JSHint output.

**Comment by nzakas** [comment]
Sorry, totally missed that. OK, we will take a look.

**Comment by tav** [comment]
Thanks @nzakas.

**Comment by nzakas** [comment]
Working on this.

**Comment by nzakas** [comment]
This was actually fixed in df9b2549438 but the issue wasn't closed due to a misconfigured comment.

**Comment by tav** [comment]
Thanks @nzakas! :+1:
